From a0a3ee2bff7dc60668448f5b70b4e7ce263f1b3b Mon Sep 17 00:00:00 2001
From: Xubao Jiang <xubaoj@codeaurora.org>
Date: Mon, 17 Aug 2020 17:13:02 +0800
Subject: [PATCH 2/2] libdp: update fs_mgr to work in LV

Update block device and dm mapper path for LV

Update fs_mgr_dm_linear.cpp to add fs_mgr_wait_for_file and
fs_mgr_get_slot_suffix support

Update base/include/android-base/logging.h to build with GCC

Change-Id: I1c3e51cc7b88d7a050acfbac22a1cc69617340d0
Signed-off-by: Xubao Jiang <xubaoj@codeaurora.org>
---
 base/include/android-base/logging.h |  4 +++
 fs_mgr/fs_mgr_dm_linear.cpp         | 61 ++++++++++++++++++++++++++++++++++++-
 fs_mgr/libdm/dm.cpp                 |  9 ++++++
 fs_mgr/liblp/partition_opener.cpp   |  4 +++
 fs_mgr/liblp/utility.cpp            | 11 ++++++-
 fs_mgr/liblp/utility.h              |  2 ++
 6 files changed, 89 insertions(+), 2 deletions(-)

diff --git a/base/include/android-base/logging.h b/base/include/android-base/logging.h
index f94cc25..30111ba 100644
--- a/base/include/android-base/logging.h
+++ b/base/include/android-base/logging.h
@@ -486,8 +486,12 @@ namespace std {
 #pragma clang diagnostic ignored "-Wgcc-compat"
 #define OSTREAM_STRING_POINTER_USAGE_WARNING \
     __attribute__((diagnose_if(true, "Unexpected logging of string pointer", "warning")))
+#if defined(__GNUC__)
+inline std::ostream& OSTREAM_STRING_POINTER_USAGE_WARNING operator<<(std::ostream& stream, const std::string* string_pointer) {
+#else
 inline std::ostream& operator<<(std::ostream& stream, const std::string* string_pointer)
     OSTREAM_STRING_POINTER_USAGE_WARNING {
+#endif
   return stream << static_cast<const void*>(string_pointer);
 }
 #pragma clang diagnostic pop
diff --git a/fs_mgr/fs_mgr_dm_linear.cpp b/fs_mgr/fs_mgr_dm_linear.cpp
index ee6ffdb..4137dab 100644
--- a/fs_mgr/fs_mgr_dm_linear.cpp
+++ b/fs_mgr/fs_mgr_dm_linear.cpp
@@ -32,6 +32,7 @@
 #include <unistd.h>
 
 #include <sstream>
+#include <thread>
 
 #include <android-base/file.h>
 #include <android-base/logging.h>
@@ -40,7 +41,21 @@
 #include <android-base/unique_fd.h>
 #include <liblp/reader.h>
 
-#include "fs_mgr_priv.h"
+#define FS_MGR_TAG "[libfs_mgr]"
+
+// Logs a message to kernel
+#define LINFO    LOG(INFO) << FS_MGR_TAG
+#define LWARNING LOG(WARNING) << FS_MGR_TAG
+#define LERROR   LOG(ERROR) << FS_MGR_TAG
+#define LFATAL LOG(FATAL) << FS_MGR_TAG
+
+// Logs a message with strerror(errno) at the end
+#define PINFO    PLOG(INFO) << FS_MGR_TAG
+#define PWARNING PLOG(WARNING) << FS_MGR_TAG
+#define PERROR   PLOG(ERROR) << FS_MGR_TAG
+#define PFATAL PLOG(FATAL) << FS_MGR_TAG
+
+enum class FileWaitMode { Exists, DoesNotExist };
 
 namespace android {
 namespace fs_mgr {
@@ -51,6 +66,46 @@ using DmTarget = android::dm::DmTarget;
 using DmTargetZero = android::dm::DmTargetZero;
 using DmTargetLinear = android::dm::DmTargetLinear;
 
+using namespace std::chrono_literals;
+
+static std::string fs_mgr_get_slot_suffix() {
+    std::string ab_suffix = "";
+    std::string cmdline;
+
+    android::base::ReadFileToString("/proc/cmdline", &cmdline);
+    for (const auto& entry : android::base::Split(android::base::Trim(cmdline), " ")) {
+        std::vector<std::string> pieces = android::base::Split(entry, "=");
+        if (pieces.size() == 2) {
+            if (pieces[0] == "androidboot.slot_suffix")
+                ab_suffix = pieces[1];
+        }
+    }
+
+    LINFO << "Get slot suffix from cmdline : " << ab_suffix;
+    return ab_suffix;
+}
+
+static bool fs_mgr_wait_for_file(const std::string& filename,
+                          const std::chrono::milliseconds relative_timeout,
+                          FileWaitMode file_wait_mode) {
+    auto start_time = std::chrono::steady_clock::now();
+
+    while (true) {
+        int rv = access(filename.c_str(), F_OK);
+        if (file_wait_mode == FileWaitMode::Exists) {
+            if (!rv || errno != ENOENT) return true;
+        } else if (file_wait_mode == FileWaitMode::DoesNotExist) {
+            if (rv && errno == ENOENT) return true;
+        }
+
+        std::this_thread::sleep_for(50ms);
+
+        auto now = std::chrono::steady_clock::now();
+        auto time_elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(now - start_time);
+        if (time_elapsed > relative_timeout) return false;
+    }
+}
+
 static bool GetPhysicalPartitionDevicePath(const LpMetadata& metadata,
                                            const LpMetadataBlockDevice& block_device,
                                            const std::string& super_device,
@@ -58,7 +113,11 @@ static bool GetPhysicalPartitionDevicePath(const LpMetadata& metadata,
     // Note: device-mapper will not accept symlinks, so we must use realpath
     // here.
     std::string name = GetBlockDevicePartitionName(block_device);
+#if defined(__linux__) && !defined(__ANDROID__)
+    std::string path = "/dev/disk/by-partlabel/" + name;
+#else
     std::string path = "/dev/block/by-name/" + name;
+#endif
     // If the super device is the source of this block device's metadata,
     // make sure we use the correct super device (and not just "super",
     // which might not exist.)
diff --git a/fs_mgr/libdm/dm.cpp b/fs_mgr/libdm/dm.cpp
index a4614d0..206aa55 100644
--- a/fs_mgr/libdm/dm.cpp
+++ b/fs_mgr/libdm/dm.cpp
@@ -22,12 +22,17 @@
 
 #include <android-base/logging.h>
 #include <android-base/macros.h>
+#include <cstring>
 
 namespace android {
 namespace dm {
 
 DeviceMapper::DeviceMapper() : fd_(-1) {
+#if defined(__linux__) && !defined(__ANDROID__)
+    fd_ = TEMP_FAILURE_RETRY(open("/dev/mapper/control", O_RDWR | O_CLOEXEC));
+#else
     fd_ = TEMP_FAILURE_RETRY(open("/dev/device-mapper", O_RDWR | O_CLOEXEC));
+#endif
     if (fd_ < 0) {
         PLOG(ERROR) << "Failed to open device-mapper";
     }
@@ -284,7 +289,11 @@ bool DeviceMapper::GetDmDevicePathByName(const std::string& name, std::string* p
     }
 
     uint32_t dev_num = minor(io.dev);
+#if defined(__linux__) && !defined(__ANDROID__)
+    *path = "/dev/dm-" + std::to_string(dev_num);
+#else
     *path = "/dev/block/dm-" + std::to_string(dev_num);
+#endif
     return true;
 }
 
diff --git a/fs_mgr/liblp/partition_opener.cpp b/fs_mgr/liblp/partition_opener.cpp
index bb8ec9c..3a0660c 100644
--- a/fs_mgr/liblp/partition_opener.cpp
+++ b/fs_mgr/liblp/partition_opener.cpp
@@ -40,7 +40,11 @@ std::string GetPartitionAbsolutePath(const std::string& path) {
     if (path[0] == '/') {
         return path;
     }
+#if defined(__linux__) && !defined(__ANDROID__)
+    return "/dev/disk/by-partlabel/" + path;
+#else
     return "/dev/block/by-name/" + path;
+#endif
 }
 
 bool GetBlockDeviceInfo(const std::string& block_device, BlockDeviceInfo* device_info) {
diff --git a/fs_mgr/liblp/utility.cpp b/fs_mgr/liblp/utility.cpp
index 72a3c57..6a47a05 100644
--- a/fs_mgr/liblp/utility.cpp
+++ b/fs_mgr/liblp/utility.cpp
@@ -25,7 +25,6 @@
 #endif
 
 #include <android-base/file.h>
-#include <ext4_utils/ext4_utils.h>
 #include <openssl/sha.h>
 
 #ifdef __ANDROID__
@@ -37,6 +36,16 @@
 namespace android {
 namespace fs_mgr {
 
+static uint64_t get_block_device_size(int fd)
+{
+    uint64_t size = 0;
+    int ret;
+    ret = ioctl(fd, BLKGETSIZE64, &size);
+    if (ret)
+        return 0;
+    return size;
+}
+
 bool GetDescriptorSize(int fd, uint64_t* size) {
 #if !defined(_WIN32)
     struct stat s;
diff --git a/fs_mgr/liblp/utility.h b/fs_mgr/liblp/utility.h
index 96f1717..262c506 100644
--- a/fs_mgr/liblp/utility.h
+++ b/fs_mgr/liblp/utility.h
@@ -26,6 +26,8 @@
 
 #include "liblp/liblp.h"
 
+#include <cstring>
+
 #define LP_TAG "[liblp]"
 #define LWARN LOG(WARNING) << LP_TAG
 #define LINFO LOG(INFO) << LP_TAG
-- 
2.7.4

